### ISSUE TITLE
Restructure Transport Parameters Section

### ISSUE THREAD

**Issue by theri** [taps-api/drafts#153]
We have Protocol Selection Properties and Path Selection Properties (§5.2 of -00), General Protocol Properties and Specific Protocol Properties (§9.1), Transport Features and Path Properties (mentioned in §9), Security Properties (§5.3), Message Properties (called Send Properties in §7.1), unless I'm forgetting some. 

- We definitely need to **restructure** the API draft.
I propose that these Properties all go in the same Section (former §9, maybe move it up in the draft more).
- Perhaps we may need to **rename** some of them, to reduce confusion, and make sure our descriptions are clear enough.
- We need to make clear **when** which of these properties can be specified and when they become immutable.

**Comment by abrunstrom** [comment]
We should also consider if Transport Parameters and Security Parameters are good names when we look over the terminology/structure.

**Comment by philsbln** [comment]
In addition to the todos raised by @theri, we also need to fix and define data types taken by the properties.
 - Most of the `boolean` typed properties are actually preferences
 - Moving the somehow-displaced definition of preferences and making this a type definition should help

Getting types/categories of parameters right might also be a first step to an IANA registry

**Comment by philsbln** [comment]
The front matter also has become tediously long as it defines a lot of things inclunding 
 - data types (cross check with #173) 
 - how preferences work
 - interaction between path- and protocol selections

We need to better separate it

**Comment by philsbln** [comment]
Rough agenda: 
1. first propose a categorization for the different kinds of parameters to the list
2. Make a PR based on the outcome

**Comment by theri** [comment]
Philipp and I just discussed the following regarding the conceptual organisation of "Properties" or “Parameters”. The goal to reduce complexity and to avoid confusion:

- Everything should be named either "Properties" or "Parameters". We should agree on one of these synonyms, not mix them. We prefer "Properties".

- Properties apply on a Connection _logical or_ a Message. So we have Connection Properties and Message Properties (see #146 for a discussion - The "Send Parameters” in the current API would then be renamed to Message Properties (and some may apply on partial messages).

- Open question: As some Properties might applicable to both a Connection and a Message, @philsbln sees the distinction between (Selection Properties/Intents/Protocol Properties) and between (Message Properties/Connection Properties) as orthogonal . @theri thinks that Selection Properties, Intents, and Protocol Properties are Connection Properties, and that Message Properties should be a category separate from them, as per-Message Path Selection is out of scope for this draft.

- Open question: It might make sense to include Security Parameters. However, we are not sure whether to rename them to "Security Properties” as this diverged from the naming in TLS.

- Properties can be either
  - Selection Properties (which are preferences for or against protocol and path features and help the Transport System choose a protocol or path),
  - Intents (which give advisory information on a communication, e.g. an expected capacity profile), or
  - Protocol Properties (which instruct the Transport System to configure a protocol once it has been chosen). [@philsbln most message properties also go here]

- Open question: Whether Intents (properties which have effects after some interpretation) are included in Selection Properties or should be a separate category.  (@theri thinks they don't need distinguishing, @philsbln wants to consider them being separate categories).

- Protocol Properties can be either generic or specific.

From this, we would derive the following categorization that are applicable to Connections:

- (Protocol and Path) Selection Properties
- Intents (may be included in Selection Properties)
- Protocol Properties
  - Generic
  - Protocol Specific
- Security Properties?

In addition, some Protocol Properties and Intents are also applicable to messages, and there are Message Properties.

Do you agree with the conceptual categorization?
Do you have other suggestions for the naming?

Once we all agree on the categorization and naming, @philsbln and @theri will prepare a Pull Request with a document structure that reflects the categorization.

**Comment by mwelzl** [comment]
That proposal sounds very good to me. I'm definitely in favor of having fewer terms!

About some of your open questions above:

> Open question: As some Properties might applicable to both a Connection and a Message, @philsbln sees the distinction between (Selection Properties/Intents/Protocol Properties) and between (Message Properties/Connection Properties) as orthogonal . @theri thinks that Selection Properties, Intents, and Protocol Properties are Connection Properties, and that Message Properties should be a category separate from them, as per-Message Path Selection is out of scope for this draft.

Just on the basis of what's written here (maybe there's a deeper reason that I'm missing?), I agree with @theri .

> Open question: Whether Intents (properties which have effects after some interpretation) are included in Selection Properties or should be a separate category. (@theri thinks they don't need distinguishing, @philsbln wants to consider them being separate categories).

Again I lean towards agreement with @theri but wonder if I'm missing something: what would be the motivation to call Intents something else - what makes them different?

**Comment by britram** [comment]
Now that IMC is done I can stop ignoring everyone else. Thanks for putting this proposal together, it's definitely moving in the right direction.

> Everything should be named either "Properties" or "Parameters". We should agree on one of these synonyms, not mix them. We prefer "Properties".

+1 to properties.

> Properties apply on a Connection logical or a Message. So we have Connection Properties and Message Properties (see #146 for a discussion - The "Send Parameters” in the current API would then be renamed to Message Properties (and some may apply on partial messages).

Yep, deferring later discussion on what the bounds of a logical connection are.

> Open question: As some Properties might applicable to both a Connection and a Message, @philsbln sees the distinction between (Selection Properties/Intents/Protocol Properties) and between (Message Properties/Connection Properties) as orthogonal . @theri thinks that Selection Properties, Intents, and Protocol Properties are Connection Properties, and that Message Properties should be a category separate from them, as per-Message Path Selection is out of scope for this draft.

I agree with @theri and @mwelzl here: the message/connection split seems to be more natural from an API client point of view. The client always knows the difference between a message and a connection, while the selection versus runtime split is somewhat more nebulous.

> Open question: It might make sense to include Security Parameters. However, we are not sure whether to rename them to "Security Properties” as this diverged from the naming in TLS.

"Security Properties" is wrong: this has a meaning in cryptographic protocol circles, and refers to the services a protocol delivers to its upper layer. "Security Parameters" is also wrong, as it has a specific meaning in TLS-adjacent contexts, and that meaning is subtly different from what these should be in a TAPS context ("security parameters" in TLS-ese is really something like "specific protocol properties" in the current document's terminology bound to TLS). Perhaps we can fix this by no longer treating security like a special snowflake, and making security-relevant protocol properties just be connection properties?

> Protocol Properties can be either generic or specific.

Open question: does this also imply that *message* properties can be generic or specific?

**Comment by philsbln** [comment]
I have thought a lot about the Properties duplicated between Message and Connection properties and it seems there is no really nice solution.

My idea why I want Message/Connection/Selection/Protocol Properties as some kind of label (and not as hierarchy) may become clearer by looking at a [Venn Diagram of Property types](https://github.com/taps-api/drafts/blob/properties/properties-venn.svg) the way I have them in mind. This assumes we want to be forward-compatible to per-message path selection, e.g., as optimization for multi-path-multi-streaming protocols like SCTP.

Still, even in this subset thingy, not all properties can be put in a single bucket.

**Comment by mwelzl** [comment]
So I think we want as much as possible to relate to a message. Some things can't, because they're only bound to time (when was the call made). Then there are mixtures, where we want to allow them to be per message because some protocols support this, but we can't guarantee.

Take Nagle. It gets activated when you make the call, and it will affect what TCP has in the buffer then. There's no way to guarantee that Nagle activation / deactivation is truly tied to a specific message. In SCTP however, message bundling is a message property.

So in conclusion it's good to allow for this to be a message property, for the transport system to try to make the best use of Nagle such that it affects the message that we asked for if possible, and for the API description to include a general disclaimer saying "Depending on the underlying protocol, it cannot always be guaranteed that message properties truly bind to a particular message."

Is this very problematic as a generic statement for all message properties? Else we need to limit it to certain properties only.

**Comment by theri** [comment]
About the Connection/Message Property split:
I agree that there's some Properties that can apply to both a Connection and a Message, as shown nicely in @philsbln's Venn diagram. Our current API draft also reflects this by duplicating "Capacity Profile" and "Instantaneous Capacity Profile" as both a Connection and a Message Property.
If we sort such properties into either category, for me this does not mean that they can not be set in the other category. For example, if "Nagle" was a Message Property, it could also be set on the Connection, e.g., as a default for all messages. We could annotate the individual Message Properties with "can also be set on the Connection", and vice versa. Of course, we would bias the application's choice of where to set it by putting the property in one category in the draft, but for me this would still be the way to go.
Alternatively, I think @philsbln's suggestion would be to have just a list of, e.g., Protocol Properties, and then label each property with "can be set on a Message", "can be set on a Connection", "can be set on both a Message and a Connection". Maybe this would suggest more flexibility to the application programmer of where to specify it.
So basically I think this is about how strongly and in what way we reflect or emphasize the Message/Property split in the document.

I understand @mwelzl's comment to be already about how to sort a property into either Message or Connection property, but I think "Nagle" can really apply to both. Is this "Immediate Acknowledgement" or something else?

> "Depending on the underlying protocol, it cannot always be guaranteed that message properties truly bind to a particular message."
> Is this very problematic as a generic statement for all message properties? Else we need to limit it to certain properties only.

I think most Message Properties do bind to a particular message and that it's pretty important to "guarantee" it by API contract, like Idempotent, Lifetime, Niceness. But yes, for other properties it makes sense to include such a statement.

> Open question: does this also imply that message properties can be generic or specific?

So far, to me all Message Properties look generic, but yes, perhaps there could be specific Message Properties as well.

**Comment by mwelzl** [comment]
Right - I thought that saying this for all is too generic. Lifetime and Niceness, well ... but Idempotence should definitely be guaranteed per message. So it's a statement that applies to some but not all properties.

About offering properties for connections and messages: I think we should simplify. Simplifying as much as we can is really important IMO.

For everything that is bound to a message, I don't see any point in *additionally* offering it per connection. In an implementation, if you want to keep the property for more than one message, just keep it set in your data structure defining all the per-message properties.

So, in the interest of simplification and text length, rather than "can be set on a Message", "can be set on a Connection", "can be set on both a Message and a Connection", I'd have per-message properties only, and then a few that apply to a connection only.

About Nagle: this is not "immediate ACK", it's coincides with what we currently call the send parameter "Immediate". It's about waiting to be able to send a larger packet vs. sending stuff out ASAP. So the "Immediate" send parameter is an example of a parameter description that would need such a disclaimer - just that it's not necessary because it's not formulated as a guarantee. This is good enough as it is.

**Comment by philsbln** [comment]
I agree with @mwelzl that for the sake of simplicity, we should not make all message properties also connection properties (which is in-line with my [venn diagram](https://github.com/taps-api/drafts/blob/properties/properties-venn.svg)).

I also see that there are some properties that come as a team: a Message Property and a Preference-Type Protocol Selection Property whether to support the Message Property. As this is a common pattern, we should make this a part of the property structure

For the case of TCP Nagle: I guess "Immediate Send" is a good message property to capture it, but might also implied by some Intents

**Comment by britram** [comment]
> For everything that is bound to a message, I don't see any point in additionally offering it per connection.

It depends on what you're offering. Is it:

- a per-connection default which message properties can override; or
- a protocol selection property that tells the connection racing machinery that it intends to send messages with that property, and that protocols guaranteeing them are preferred (or, in the case of oddly strict application developers, required). 

If it's the former, I agree with @mwelzl that this doesn't belong in the core API (though I can see most programming environments atop TAPS providing a way to bind default send properties to a connection as syntactic sugar, just to make the API less scary to its users). The latter is something I really think we do need (although very friendly programming environments could also provide static code analyzers to fill out protocol selection properties based on the message properties used).

**Comment by theri** [comment]
It is the former, and I can live with not having that in the API document, as long as nothing prevents an implementation from binding default Message Properties (not send ;) ) to a Connection.

Protocol Selection Properties, for me, clearly apply only to (future) Connections, not to Messages.

**Comment by philsbln** [comment]
Waiting for #181 to land before making a PR as this involves moving a lot of text

**Comment by britram** [comment]
#181 merged, this can go ahead now yes? (I'll also remove the discuss label, since we should hold that for stuff we want on the WG agenda for now)

**Comment by philsbln** [comment]
Started working on it in the properties branch – @britram you can take a look at it if you like.

Turned out that the restructuring made a serious amount of inconsistencies visible. Some we could fix, some that will need discussion. We should have a PR by Tuesday, but it will still have a serious amount of visible problems left (what were in the document before, but hard to spot)

**Comment by mwelzl** [comment]
Hi,

I'd indeed appreciate if this would happen by Tuesday - I was planning to do some work this week and I'm delaying it because it probably depends on this PR - e.g. issue #159
